cli/loader_lib.c

```diff
@@ -153,6 +153,177 @@ JL_DLLEXPORT const char * jl_get_libdir()
+// which can break loading of some system libraries: <https://github.com/JuliaLang/julia/issues/34276>.
+
+// Return the path to the libstdcxx to load. Otherwise, print the error and exit.
+// The path returned must be freed.
+static char* libstdcxxprobe(void)
+{
+    // Create the pipe and child process.
+    int fork_pipe[2];
+    int ret = pipe(fork_pipe);
+    if (ret == -1) {
+        fputs("Error during libstdcxxprobe:", stderr);
+        perror("pipe");
+        fflush(stderr);
+        exit(1);
+    }
+    // Make sure the buffers are clean before forking.
+    fflush(stdout);
+    fflush(stderr);
+    pid_t pid = fork();
+    if (pid == -1)  {
+        fputs("Error during libstdcxxprobe:", stderr);
+        perror("fork");
+        fflush(stderr);
+        exit(1);
+    }
+    if (pid == (pid_t) 0) { // Child process.
+        ret = close(fork_pipe[0]);
+        if (ret == -1) {
+            fputs("Error during libstdcxxprobe in child process:", stderr);
+            perror("close");
+            fflush(stderr);
+            _exit(1);
+        }
+
+        // Open the first available libstdc++.so.
+        // If it can't be found, report so by exiting zero.
+        void *handle = dlopen("libstdc++.so", RTLD_LAZY);
+        if (!handle) _exit(0);
+
+        // See if the version is compatible
+        char *dlerr = dlerror(); // clear out dlerror
+        void *sym = dlsym(handle, GLIBCXX_LEAST_VERSION);
+        dlerr = dlerror();
+        if (dlerr) {
+            // We can't use the library that was found, use the default one instead.
+            char * libpath = JL_DEFAULT_STDCXXLIB;
+            size_t written_sofar = 0;
+            size_t write_len = strlen(libpath);
+            while(write_len) {
+                ssize_t bytes_written = write(fork_pipe[1], libpath + written_sofar, write_len);
+                if (bytes_written == -1 && errno == EINTR) continue;
+                if (bytes_written == -1 && errno != EINTR) {
+                    fputs("Error during libstdcxxprobe in child process:", stderr);
+                    perror("write");
+                    fflush(stderr);
+                    _exit(1);
+                }
+                write_len -= bytes_written;
+                written_sofar += bytes_written;
+            }
+            _exit(0);
+        }
+
+        // No error means the symbol was found, we can use this library.
+        // Get the path to it.
+        struct link_map * lm;
+        ret = dlinfo(handle, RTLD_DI_LINKMAP, &lm);
+        if (ret == -1) {
+            fputs("Error during libstdcxxprobe in child process:", stderr);
+            fprintf(stderr, "dlinfo - %s\n", dlerror());
+            fflush(stderr);
+            _exit(1);
+        }
+        char * libpath = lm->l_name;
+
+        // Write the path over the pipe to the parent process.
+        size_t written_sofar = 0;
+        size_t write_len = strlen(libpath);
+        while(write_len) {
```

Comment:
style nit: while is not a function call
```suggestion
        while (write_len) {
```